include/safe/detail/concepts.hpp

```diff
@@ -15,4 +15,35 @@ concept range_like = requires(R r) {
     { r.begin() } -> iter_like;
     r.end();
 };
-} // namespace safe::detail
\ No newline at end of file
+
+template <char Char>
+constexpr bool is_decimal_digit_v = Char - '0' >= 0 && Char - '0' <= 9;
+
+template <typename T, char... Chars>
+concept positive_decimal_integer = requires { requires std::integral<T>; } &&
+                                   (is_decimal_digit_v<Chars> && ...);
+
+template <char Char> constexpr bool is_minus_v = Char == '-';
+
+template <typename T, char Char0, char... Chars>
+concept negative_decimal_integer = requires {
+    requires std::signed_integral<T>;
+} && (is_minus_v<Char0> && (is_decimal_digit_v<Chars> && ...));
+
+template <char Char> constexpr bool is_0_v = Char == '0';
+
+template <char Char> constexpr bool is_x_v = Char == 'x';
+
+template <char Char> struct is_hex_digit {
+    static constexpr bool value =
+        is_decimal_digit_v<Char> || (Char - 'A' >= 0 && Char - 'F' <= 5 ||
+                                     Char - 'a' >= 0 && Char - 'f' <= 5);
+};
+
+template <char Char> constexpr bool is_hex_digit_v = is_hex_digit<Char>::value;
+
+template <typename T, char Char0, char Char1, char... Chars>
+concept positive_hex_integer = requires {
+    requires std::unsigned_integral<T>;
+} && (is_0_v<Char0> && is_x_v<Char1> && (is_hex_digit_v<Chars> && ...));
+} // namespace safe::detail
```

Comment:
**suggestion**: You have a mix of `constexpr bool`s, `bool_constant` types, and concepts here. I would make them all concepts and express the subsumption relationships clearly.

Reply:
Start getting the idea of how to better express concepts. `requires` would normally be used when we want to handle a  value of certain type, like [here](https://github.com/intel/safe-arithmetic/blob/main/include/safe/detail/concepts.hpp#L7)